language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs

```diff
@@ -92,3 +95,132 @@ getBuiltinFix = do
           ]
+
+-- | A type that looks a bit transformation.
+--
+-- > trans F G Q : F Q -> G Q
+trans :: Type TyName () -> Type TyName () -> Type TyName () -> Quote (Type TyName ())
+trans f g q = pure $ TyFun () (TyApp () f q) (TyApp () g q)
+
+-- | A type that looks like a natural transformation.
+--
+-- > natTrans F G : forall Q :: * . F Q -> G Q
+natTrans :: Type TyName () -> Type TyName () -> Quote (Type TyName ())
+natTrans f g = freshTyName () "Q" >>= \q -> TyForall () q (Type ()) <$> trans f g (TyVar () q)
+
+-- | A type that looks like a natural transformation to Id.
+--
+-- > natTransId F : forall Q :: * . F Q -> Q
+natTransId :: Type TyName () -> Quote (Type TyName ())
+natTransId f = do
+    q <- freshTyName () "Q"
+    pure $ TyForall () q (Type ()) (TyFun () (TyApp () f (TyVar () q)) (TyVar () q))
+
+-- | The 'fixBy' combinator.
+--
+-- fixBy :
+--     forall F :: (* -> *) .
+--     ((forall Q :: * . F Q -> Q) -> (forall Q :: * . F Q -> Q)) ->
+--     ((forall Q :: * . F Q -> F Q) -> (forall Q :: * . F Q -> Q))
+getBuiltinFixBy :: Quote (Term TyName Name ())
+getBuiltinFixBy = do
+    f <- freshTyName () "F"
+    by <- freshName () "by"
+    byTy <- do
+        nt1 <- natTransId (TyVar () f)
+        nt2 <- natTransId (TyVar () f)
+        pure $ TyFun () nt1 nt2
+
+    instantiatedFix <- do
+        fix <- getBuiltinFix
+        nt1 <- natTrans (TyVar () f) (TyVar () f)
+        nt2 <- natTransId (TyVar () f)
+        pure $ TyInst () (TyInst () fix nt1) nt2
+
+    recc <- freshName () "rec"
+    reccTy <- do
+      nt <- natTrans (TyVar () f) (TyVar () f)
+      nt2 <- natTransId (TyVar () f)
+      pure $ TyFun () nt nt2
+
+    h <- freshName () "h"
+    hty <- natTrans (TyVar () f) (TyVar () f)
+
+    r <- freshTyName () "R"
+    fr <- freshName () "fr"
+    let frTy = TyApp () (TyVar () f) (TyVar () r)
+
+    q <- freshTyName () "Q"
+    fq <- freshName () "fq"
+    let fqTy = TyApp () (TyVar () f) (TyVar () q)
+    pure $
+        TyAbs () f (KindArrow () (Type ()) (Type ())) $
+        LamAbs () by byTy $
+        Apply () instantiatedFix $
+        LamAbs () recc reccTy $
+        LamAbs () h hty $
+        TyAbs () r (Type ()) $
+        LamAbs () fr frTy $
+        Apply () (TyInst () (
+                         Apply () (Var () by) $
+                          TyAbs () q (Type ()) $
+                          LamAbs () fq fqTy $
+                          Apply () (TyInst () (Apply () (Var () recc) (Var () h)) (TyVar () q)) $
+                          Apply () (TyInst () (Var () h) (TyVar () q)) (Var () fq))
+                     (TyVar () r))
+        (Var () fr)
+
+-- | Make a @n@-ary fixpoint combinator.
+--
+-- > getBuiltinFixN n :
+--       forall A1 B1 ... An Bn :: * .
+--       forall Q :: * .
+--           ((A1 -> B1) -> ... -> (An -> Bn) -> Q) ->
+--           (A1 -> B1) ->
+--           ... ->
+--           (An -> Bn) ->
+--           Q
+--       forall R :: * . (A1 -> B1) -> ... (An -> Bn) -> R
```

Comment:
It should be `forall R :: * . ((A1 -> B1) -> ... (An -> Bn) -> R) -> R`, I think.